### DISCUSSION TITLE
What's the best way to share code between API routes and the app itself?

### DISCUSSION THREAD

**Discussion by selbekk** [vercel/next.js#17890] [Help]
Just wondering if there exists some kind of best practice on how to share code between API routes and the rest of the nextjs app? Should I have a `shared` folder? A `domain` folder? Where should I put my stuff??!?

**Comment by balazsorban44** [comment]
What kind of code are you taking about? Data fetching? Auth? Something else?

In any case, I don't think Next.js has any strong opinions. the only restraints are the public and pages folder, either in the root or in an src folder. 

We have a components folder for shared/reusable React components, a utils for plain JS functions, a lib for setting up third party stuff, and two files that contain a bunch of data fetcher functions for graphql separate for server and client side, for code splitting.

We use react-query together with my custom graphql client solution, which I am planning to share in the near future with the public. no runtime graphql, it's embarrassingly simple, but could not find anything like it. 😳

**Reply by selbekk** [reply]
Well, to start out it's shared enums and type definitions, but I assume there will be quite a bit of shared code (data fetching and authentication comes to mind). 

Would love to hear other ways people have structured this!

Same goes for server code, if it's too big to fit into an API route file

**Reply by balazsorban44** [reply]
I put shared types in a root folder called `@types`, and add it to the `tsconfig.json`:

```js
{
  "compilerOptions": {
   //...
    "typeRoots": [
      "node_modules/@types",
      "@types",
    ],
  },
  "exclude": [
    "node_modules"
  ],
}
```

For auth, I use `next-auth`, that makes it almost unnecessary to write any auth related code, supporting a bunch of OAuth providers by default: https://next-auth.js.org/configuration/providers

So the last remaining part is data fetching. We fetch most of the data in `getStaticProps` functions. Those import a simple data fetcher function from a file. They look like this:

```js
export const getFrontpage = async (client, variables) => {
  const result = await client.request(FrontpageQuery, variables)
  const layout = await getLayoutData() // similar function to `getFrontpage`, but fetches common layout info like menu, footer etc. Next.js currently does not support `getStaticProps` in `_app`, so this must be fetched on each page separatly.
  if (!result?.page.frontpage) {
    throw {
      hasFrontPageData: !!result?.page.frontpage,
      message: "Missing frontpage data",
    }
  }
  return {
    frontpage: result.page.frontpage,
    ...layout,
  }
}
```

These functions take a client that communicates with the data source, and variables (we use GraphQL).

This pattern is nice because it can be reused on the frontend and the backend, by sending a different client to it, and these functions have the job to fetch, error check, and map the data if necessary. Caching etc. happens with `react-query`

<details>
<summary>(opinionated) structure example</summary>
<img src="https://user-images.githubusercontent.com/18369201/96115360-1b1c3600-0ee7-11eb-8af9-9d1e093fbbfc.png"/>
</details>

**Reply by mikedelgaudio** [reply]
Also ran into this question, I am fairly new but I'd like to share an error handler between different API routes. I wasn't sure if there was a best practice on this.

**Comment by bfunc** [comment]
Any clue? Looks like Next.js splits each API route into its own bundle, so single instance inside module scope will not work, because module is re-created when user enters another route.

So far only solution that works for me (sharing levelDB instance between API routes) is to use global namespace object. Like this:
```javascript
// lib/db.js
import level from 'level'

export const getDB = () => {
  if (!global.database) {
    // save DB instance in global object for reuse
    global.database = level('__db__')
  }
  const db = global.database

  async function findById(id) {
    const result = await db.get(id)
    return JSON.parse(result)
  }

  async function find({ email } = {}) {/* ... */}
  async function update(id, payload) {/* ... */}
  async function create(payload) {/* ... */}

  return {
    find,
    findById,
    update,
    create,
  }
}
```

**Reply by GustavPS** [reply]
Did you ever find a better solution for this? Facing the same issue and I really don't want to be forced to use a global object just to share an instance

**Reply by Joker666** [reply]
Same issue for me

**Comment by amie-wilt** [comment]
Has anyone found a better solution for this?

**Reply by GustavPS** [reply]
Sadly no, I'm currently forced to use globals. Please let me know if you find a better solution

**Comment by justin-together** [comment]
I'd like to share code between an API route and a WebSocket proxy that I believe can only be implemented with a custom server.
